### ISSUE TITLE
Support global vulnerability analysis policies

### ISSUE THREAD

**Issue by nscuro** [DependencyTrack/hyades#930]
### Context

Since v4.x, Dependency-Track no longer supports global auditing of vulnerabilities. 

This was done because analyses are not generally applicable to all projects depending on a given component. Context matters, and decisions like "not affected" can only ever be made on the project level.

It is thus also not possible to suppress false positives for all projects in a portfolio anymore. However, the community still has a need for it: https://github.com/DependencyTrack/dependency-track/issues/1495

Automating analysis works, utilizing both Webhook notifications and Dependency-Track's REST API. An example implementation of such automation is available with [dtapac](https://github.com/nscuro/dtapac).

What's problematic about this, is that at the point at which Webhooks are sent, it's already too late:

* Users (and downstream systems) are already notified about a new vulnerability having popped up
* CI systems already pulled findings via REST API, potentially failing a quality gate

Projects like OWASP Dependency-Check support [suppression files](https://jeremylong.github.io/DependencyCheck/general/suppression.html). Typically based on regular expressions for component identifiers, and vulnerability IDs.

Dependency-Track supports more than just suppression though.

### How does this differ from VEX / VDR?

While VEX and VDR can be used to communicate analysis of a given finding (which may lead to suppression), the analysis in both cases is static and can not be bound to additional constraints.

### Why not use the existing policy feature?

Policy evaluation happens "too late" in the workflow. When the evaluation is executed, findings have already been created, and `NEW_VULNERABILITY` notification have already been sent. This is sub-optimal, as users are notified about findings that ultimately may be suppressed anyway. For CI use cases, build systems could fetch the list of findings before policy evaluation completed.

### Proposed solution

Introduce a new type of policy, entirely dedicated to applying analyses in an automated manner.

Policies consist of the following:

* Metadata (name, description, created, updated dates, ...)
* One or more conditions, provided as CEL expressions
  * Expressions may use everything that's already available in [CEL policies](https://dependencytrack.github.io/hyades/latest/usage/policy-compliance/expressions/)
* An analysis that shall be applied when conditions match

Example policy (represented as YAML for brevity):

```yaml
name: Example
description: Foo bar
author: Jane Doe
created: 2023-11-22T06:06:06Z
updated: 2023-11-23T07:07:07Z
validUntil: 2024-01-01T00:00:00Z
conditions:
  - |-
   component.name == "foo" 
    && project.name == "bar"
    && "internal" in project.tags
    && !component.is_dependency_of(org.dependencytrack.policy.v1.Component{group: "org.springframework.boot"}
analysis:
  state: NOT_AFFECTED
  justification: CODE_NOT_REACHABLE
  details: Because foo bar baz
  suppress: true
```

The following describes a potential workflow. Key components:

* `VulnScanResultProcessor`: The Kafka consumer that processes [`ScanResult`s ](https://github.com/DependencyTrack/hyades/blob/418da1d20e38a849b28c34b0f3b54e785529fbda/proto/src/main/proto/org/hyades/vulnanalysis/v1/vuln_analysis.proto#L71-L80)
* `AnalysisPolicyProvider`: An abstraction for managing (creating, updating, retrieving) policies
  * Providers may have various implementations, ranging from hardcoded policies, policies stored in Git / S3, etc.
  * Providers may optionally perform additional filtering to decide which policy is applicable to which component/project
  * We can potentially allow for custom Plugins by offering an SPI
* `AnalysisPolicyEvaluator`: A service that evaluates policies, given the context of:
  * A vulnerability
  * The affected component
  * The project the component belongs to

```mermaid
sequenceDiagram
    Kafka->>+VulnScanResultProcessor: Vuln Scan Result
    VulnScanResultProcessor->>VulnScanResultProcessor: Parse
    loop for each vulnerability
        VulnScanResultProcessor->>Database: Sync vuln details
        VulnScanResultProcessor->>AnalysisPolicyEvaluator: Evaluate policies
        Note over VulnScanResultProcessor, AnalysisPolicyEvaluator: Context: Component, Project, Vulnerability
        AnalysisPolicyEvaluator->>AnalysisPolicyProvider: Get applicable policies
        Note over AnalysisPolicyEvaluator, AnalysisPolicyProvider: Provider could be SPI to plug in various<br/>backends (Git, S3, Database, ...)
        AnalysisPolicyProvider->>AnalysisPolicyEvaluator: Applicable policies
        AnalysisPolicyEvaluator->>AnalysisPolicyEvaluator: Evaluate policies
        AnalysisPolicyEvaluator->>VulnScanResultProcessor: state=NOT_AFFECTED<br/>justification=CODE_NOT_REACHABLE<br/>detail=Because foo bar baz<br/>suppress=true
        Note over AnalysisPolicyEvaluator, VulnScanResultProcessor: What if multiple policies match?<br/>Short-circuit, prioritization?
        critical atomically
            VulnScanResultProcessor->>Database: Create finding
            VulnScanResultProcessor->>Database: Auto-apply analysis
        end 
        opt finding not suppressed
            VulnScanResultProcessor->>Kafka: Send NEW_VULNERABILITY<br/>notification 
        end
    end
```

When an analysis is applied, it will update the finding's audit trail. Just as if it was done manually via the Dependency-Track UI.

**Comment by nscuro** [comment]
* Additional field: `validFrom` to track the start date of when a policy is valid
* How do we know when a policy expired (validUntil)? Do we need to record this somehow?
* Likely need a new database table for vulnerability policies (name not set in stone)
* Users should be able to create, update, delete policies via UI still if they want
  * Enabling sync from external source will set those policies to read only
* Should be able to capture user who created / updated policy in DT based on `Principal`
* This feature should be behind a feature flag at first :)
* JSON schema is needed to validate policy YAML files

**Comment by nscuro** [comment]
- [ ] Define a JSON schema that can be used to validate policies in YAML format against
  - The schema will be used both in CI, *and* by DT during policy reconciliation
  - Should be as strict as possible, using formats (e.g. `date-time`) and regex validation wherever possible
  - Consider enforcing a mandatory `apiVersion` field, to make schema evolution easier
- [ ] Define a new database table schema for vulnerability policies
- [ ] Define and implement policy reconciliation, which should include:
  - Fetching of policy bundles from blob storage / file server
    - Schedule and endpoint must be configurable
  - Extraction of the bundle
  - Validation of all included policy files
  - Synchronization with database (single DB transaction, this needs to be atomic)
  - Appropriate logging of errors, such that they can be monitored for and reacted to
- [ ] Implement `VulnerabilityPolicyEvaluator`
  - Either as new class, or integrate into existing `CelPolicyEngine`
  - Can develop against a `VulnerabilityPolicyProvider` interface at first, which can easily be mocked
  - Same as `CelPolicyEngine`, it will potentially need to load additional data from DB based on condition expression requirements
  - Think about where caching may be used to avoid redundant expensive evaluation
- [ ] Integrate `VulnerabilityPolicyEvaluator` in [`VulnerabilityScanResultProcessor`](https://github.com/DependencyTrack/hyades-apiserver/blob/main/src/main/java/org/dependencytrack/event/kafka/processor/VulnerabilityScanResultProcessor.java) as outlined above
  - Add Micrometer `Timer` to track time taken for evaluation
  - Consider logging warnings when evaluation takes too long
  - Profile the integration, it's crucial that it performs well!

**Comment by nscuro** [comment]
@valentijnscholten Good suggestion. The way we did it for the existing policy functionality, is that users can continue to use the pre-defined conditions (e.g. `Coordinates`), but in the background we re-write them to CEL and execute them that way.

Offering a set of condition templates as you say also makes sense to me.

**Comment by tomaszn** [comment]
@nscuro This is fabulous.

One thing seems missing to me. Can the scope of the rule updates be limited per user? For example, I would like to limit users having access only to projects A and B to be only permitted to create policies that cover only these projects. This check would apply to changes both via UI, and also via API, so separate bot accounts could be used per team, to prevent one team from suppressing findings in other team's projects. This would be useful in the scenario of policy management in Git with synchronization via REST API, where each team or project could maintain their policies and update them e.g. in CI just before uploading the SBOM.

**Comment by skhokhlov** [comment]
I'm so excited about this feature. One thing I want to suggest is to have a policies archive push endpoint in api and ui. In this case policies can be properly controlled and DT won't need any integration with file storage systems. Also it allows to rollback policies to a previous version without any troubles. In case of polling an external endpoint for policies archive, the problem may be that artefact storage won't allow to override file with a new version. It means that for each new version the link will be different.

**Comment by rkg-mm** [comment]
This would address https://github.com/DependencyTrack/dependency-track/issues/2183, and I totally need this yesterday 😆 

One idea: In the existing vulnerability analysis view, add a button like "Apply to all project versions", clicking it opens the UI to create a corresponding policy, fields pre-filled and customizable if the user wants to. This would make the workflow much easier (teams usually work in the vulnerability view, do an analysis starting from there and add the result in this view. Forcing them to then switch to another view entering details that already were there in the view before would be a usability nightmare for this use case)

**Comment by nscuro** [comment]
> @rkg-mm One idea: In the existing vulnerability analysis view, add a button like "Apply to all project versions", clicking it opens the UI to create a corresponding policy, fields pre-filled and customizable if the user wants to.

I like this idea!

> However, creating policies affecting a specific project (can it include children?) could be done if the vulnerability analysis permission is there for the user and this project.

I think this is an ACL check we should be able to do. At least off the top of my head I don't see anything that would prevent us from doing that. *But*, this and the idea you mentioned earlier necessitate some form of policy "ownership", which likely needs to be scoped to teams. So related to what @tomaszn mentioned.

When users create policies from the audit view, the policies are owned by the team(s) that user is a member of.

**Comment by nscuro** [comment]
The feature has been implemented for a while, and used successfully in a large-scale deployment. It previously required a GitOps-like approach, as outlined in the issue description.

I am currently in the process of adding the capability to manage these policies via UI and / or REST API: https://dependencytrack.github.io/hyades/snapshot/architecture/decisions/016-user-managed-vuln-policies/

Once that is done, this issue can be considered done.

**Comment by rkg-mm** [comment]
> Once that is done, this issue can be considered done.

Hmm unfortunately this does not solve the use case where one project tracking n-versions of their project want to apply an assessment to all versions of this project, or did I misunderstand your concept? Would you consider this a separate feature - and if yes is there a request for that already that you know of?

**Comment by nscuro** [comment]
@rkg-mm You will be able to achieve this by either:

* Binding the policy to a tag, and reusing that tag across all your project versions
* Keeping the policy "global" (i.e. not binding it to a tag), and encoding this into the CEL condition instead, e.g.:
   ```js
   project.name in ["foo", "bar"] // Applies to all versions of project foo and bar
   ```

**Comment by rkg-mm** [comment]
In theory: yes. Practically: No.
Reason:
Permissions and responsibilities ;-)

The analysis is done by each project team for their own projects. If I understand your concept, I would need to grant everybody the permission  to modify these policies - which I cannot do as people could affect projects they are not responsible for, not even see who uses a tag since everybody only sees their own projects. 
From my point of view your concept works for a global team that assesses policies for all projects, but not for projects working under own responsibilities.

Instead it would require a "enroll assessment for this vulnerability to all versions of this project" button, which is scoped to the current project only.

**Comment by nscuro** [comment]
@rkg-mm I think the use case you're describing doesn't benefit from vulnerability policies. Fundamentally these policies are meant to be a governance tool, where you can make broadly applicable decisions based on (potentially complex) conditions.

In your case, you'd use policies when multiple of your teams keep seeing the same false positives across many projects, or when you have risk acceptance agreements that allow you to ignore findings in specific contexts. You'd then add a policy which would help all your teams at once, without them having to take any action.

For the "enroll assessment for this vulnerability to all versions of this project" use case, vulnerability policies would be overkill IMO. Presumably you don't need dynamic conditions, but you want a fixed assessment (e.g. suppression) to be applied to the same component <-> vuln combination across all versions of a project. 

With the current model, we could implement that via "bulk analysis". Say we add a "apply to all project versions" checkbox to the audit view. The limitation is that it would only work for versions we *know of*, i.e. those that already exist. For the "... and all future versions" angle we really would need #1736, which would allow us to associate an assessment with the overarching project, instead of a specific project version.

Is my understanding of your request correct? And if so, how critical is the "... and all future versions" part for you, outside of it obviously being super useful?

**Comment by nscuro** [comment]
I guess an extended question would be: If you say "apply this assessment to all versions of the project", how does that translate to component identity? Presumably you'd update component versions between project versions, which changes the identity of the component, and the assessment couldn't be applied as-is.

**Comment by rkg-mm** [comment]
>Is my understanding of your request correct? And if so, how critical is the "... and all future versions" part for you, outside of it obviously being super useful?

Yes you got that correct. And I agree, this is likely another use case. Do we have a feature request for this already?

All known versions would be fine for me for the moment. We anyway usually clone new versions from the last version in pipeline before we upload the new SBOM, including all assessments, which should cover the future version then with the same assessment, too.

>I guess an extended question would be: If you say "apply this assessment to all versions of the project", how does that translate to component identity? Presumably you'd update component versions between project versions, which changes the identity of the component, and the assessment couldn't be applied as-is.

Since its the same vulnerability, applying to multiple versions of the same component, I'd say the assessment should be applied to all occurances of this vulnerability to components with the same identifier, regardless the version. The typical use case is: Component exists in multiple versions in multiple project versions. Vulnerability hits half of them. If the usage of the component never changed, just the versions updated, you want to apply this to all of them. 
If, however, the vulnerability also applies to a totally different component, then I would suggest not to apply this to the other component, as you should verify each component usage independently.

**Comment by nscuro** [comment]
> @rkg-mm Do we have a feature request for this already?

I think this might come closest? https://github.com/DependencyTrack/dependency-track/issues/2107

> Since its the same vulnerability, applying to multiple versions of the same component, I'd say the assessment should be applied to all occurances of this vulnerability to components with the same identifier, regardless the version.

We'll need to be very careful WRT what "same identifier" actually means. i.e. `pkg:npm/acme-lib@1.2.3` is technically different than `pkg:npm/acme-lib@1.2.3?foo=bar`. Similar issue with components sharing the same PURL but having different CPEs.

**Comment by rkg-mm** [comment]
Related but talking about different projects, probably not only different versions.

Hmm... what could be a good solution? Offering the user a list of affected "candidates" grouped by version, in which he could uncheck the ones he doesn't want to update? I assume in most cases this is not relevant and you apply it to all, but I understand your point that we should be technically correct somehow.

**Comment by nscuro** [comment]
> Related but talking about different projects, probably not only different versions.

Could you then please raise a new issue with your specific requirements?

It would also be useful to research how others in the space are handling these things, and whether those solutions actually worked out to be good. Sonatype's Lifecycle product has a [Waivers](https://help.sonatype.com/en/waivers.html) concept which looks related. Based on [this](https://help.sonatype.com/en/examples-of-waiver-scoping.html), it seems Lifecycle Waivers are broadly scoped to component coordinates (e.g. group, name, version).

**Comment by rkg-mm** [comment]
will do. in the main repo or in hyades repo?

**Comment by nscuro** [comment]
Let's do main, we'll move everything over from hyades for GA anyway. Also the main repo has likely more eyes on it.

**Comment by rkg-mm** [comment]
@nscuro I found an existing feature request for this second case - funnily I created this myself 4 years ago :-D. https://github.com/DependencyTrack/dependency-track/issues/2183
I updated the description with some points of our discussion here.
